### DISCUSSION TITLE
Installation doesn not work because there is no tar command

### DISCUSSION THREAD

**Discussion by dotvav** [Arkady23/V380-Web#58] [Q&A] [answered]
The setup script does not work properly as the `tar` command is nowhere to be found on my firmware. Apart from untaring the archive prior and altering the setup script, what can I do? Can I install tar afterwards?

**Comment by Arkady23** [comment]
Can you give me a screenshot of the installation and the installation log for understanding the situation?

**Comment by dotvav** [comment]
Thanks, Arkady.

Camera model : XiaoviXVV 1080p model MVR1120S-B1. NB: I have disassembled it, connected UART, and I have access to the bootloader and linux through serial, if needed.

Starting from scratch, I unzip the downloaded files on an empty sdcard.
```
Microsoft Windows [version 10.0.22000.1936]
(c) Microsoft Corporation. Tous droits réservés.

C:\Users\dotvav>d:

D:\>dir
 Le volume dans le lecteur D n’a pas de nom.
 Le numéro de série du volume est 43D2-8DEF

 Répertoire de D:\

29/05/2023  07:04    <DIR>          ark-add-on
29/05/2023  07:04                13 quick_check.ini
29/05/2023  07:04               319 setup.sh
               2 fichier(s)              332 octets
               1 Rép(s)  32 022 904 832 octets libres

D:\>cd ark-add-on

D:\ark-add-on>dir
 Le volume dans le lecteur D n’a pas de nom.
 Le numéro de série du volume est 43D2-8DEF

 Répertoire de D:\ark-add-on

29/05/2023  07:04    <DIR>          .
29/05/2023  07:04    <DIR>          ..
29/05/2023  07:04         1 562 624 install_20220927
29/05/2023  07:04               171 startup.sh
29/05/2023  07:04               353 update.sh
               3 fichier(s)        1 563 148 octets
               2 Rép(s)  32 022 904 832 octets libres

D:\ark-add-on>
```

I unmount the sdcard from my PC, put it in the cam, plug the power into the cam. 

Then I telnet into the cam's IP address.
```

(none) login: root
~ # cd /mnt/sdcard
/mnt/sdcard # ls -al
drwx------ root     root              2023-05-29 05:22 RecFiles
drwx------ root     root              2023-01-24 09:37 System Volume Information
drwx------ root     root              2023-05-29 07:04 ark-add-on
-rwx------ root     root           13 2023-05-29 07:04 quick_check.ini
-rwx------ root     root         2057 1970-01-01 00:00 quick_check_dump_2594XXXX.txt
-rwx------ root     root          319 2023-05-29 07:04 setup.sh
/mnt/sdcard # ./setup.sh
/mnt/sdcard # tar
-sh: tar: not found
/mnt/sdcard # busybox
BusyBox v1.24.1 (2017-09-29 00:55:28 PDT) multi-call binary.
BusyBox is copyrighted by many authors between 1998-2015.
Licensed under GPLv2. See source distribution for detailed
copyright notices.

Usage: busybox [function [arguments]...]
   or: busybox --list
   or: function [arguments]...

        BusyBox is a multi-call binary that combines many common Unix
        utilities into a single executable.  Most people will create a
        link to busybox for each function they wish to use and BusyBox
        will act like whatever it was invoked as.

Currently defined functions:
        ash, awk, cat, chmod, clear, cp, date, dd, devmem, df, dhcprelay, dmesg, dnsdomainname, dumpleases, echo, egrep, env, fgrep, find, flock,
        free, ftpd, getty, grep, halt, head, hostname, ifconfig, inetd, init, insmod, ipcs, kill, killall, killall5, linuxrc, ln, logger, login,
        logread, ls, lsmod, lsusb, md5sum, mkdir, mkdosfs, mkfs.vfat, modinfo, more, mount, mv, netstat, passwd, ping, pkill, poweroff, printf, ps,
        pwd, reboot, renice, reset, rm, rmmod, route, sed, sh, sleep, sort, strings, sync, sysctl, syslogd, tail, telnet, telnetd, tftp, time, top,
        touch, traceroute, tty, udhcpc, udhcpd, umount, uname, unlink, uptime, usleep, vi, volname, watch, wc
MVX2####g9dfae01CM_BYBOX1241#XVM

```

You can see that `tar` gives `-sh: tar: not found` and it is not present in busybox. 

```
/mnt/sdcard # cat /mnt/mtd/vg_boot.sh
#!/bin/sh



sd=/mnt/sdcard/ark-add-on
if [ -f "$sd/startup.sh" ]; then
        $sd/startup.sh &
fi
/mnt/sdcard # 
```

The `setup.sh` script has done its job. But `update.sh` is failing at `tar`. This is why I am going to extract the tarball manually in the sd card, and remove the `tar` commands from `update.sh`.

In case it is of any use, here is the content of the quick_check dump:
```
===== quick check result dump =====
devID: 2594XXXX
====== sensor ====>
/tmp/sensorid_GC2063_MIPI
======version ====>
[app_version]
name=AppMSCE1_V2.3.0.2
date=20190416
[kernal_version]
name=KerMSCE1_IPC38-UART_V3.3.2
date=20180817
[hardware_version]
name=HwMSCE1_WF3_IPC38-UART
[base_version]
name=AppMSCE1_V2.2.1.1
date=20190416
======cParams ====>
[CONST_PARAM]
ircut_flip=0
color_flip=0
firmware_type=23
network_reset_mode=1
vframe_encode_type=1
aframe_encode_type=1
uart2_usage=0
lens_type=0
pano_auto_lens_type=0
vframe_fr_nolimit=0
dev_i2c=/dev/i2c-gpio
reversal_value=1
starlight_mode=0
light_ctrl=0
night_with_color=0
ptz_opposite_run=3
rtsp=1

[FRAME_QUANT_REC]
main_qp_max=51
main_qp_intra=39
main_qp_min=38
sub_qp_max=43
sub_qp_intra=39
sub_qp_min=38

[FRAME_QUANT_MOBILE]
main_qp_max=51
main_qp_intra=39
main_qp_min=38
sub_qp_max=43
sub_qp_intra=39
sub_qp_min=38

[FRAME_QUANT_ONVIF]
main_qp_max=51
main_qp_intra=39
main_qp_min=38
sub_qp_max=43
sub_qp_intra=39
sub_qp_min=38

[PTZ_VER]
vague_full_ag_H=0
vague_full_ag_V=0
has_no_limiter=1
======usb ====>
Bus 002 Device 002: ID 0bda:f179
Bus 001 Device 001: ID 1d6b:0002
Bus 002 Device 001: ID 1d6b:0002
======network ====>
eth0      Link encap:Ethernet  HWaddr 00:30:1B:XX:XX:XX  
          BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
          Interrupt:90 Base address:0x2000 

lo        Link encap:Local Loopback  
          inet addr:127.0.0.1  Mask:255.0.0.0
          UP LOOPBACK RUNNING  MTU:65536  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0 
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)

nameserver 1.1.1.1
nameserver 1.0.0.1
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
```

**Reply by Arkady23** [reply]
Thank you for the detailed information.
I think you have a truncated Linux. Therefore, it is unlikely that you will succeed. I recently ordered a new camera after I realized that the new cameras have truncated Linux to try experimenting. But they sent me a camera with full Linux. Therefore, the problem remained unresolved.

**Comment by dotvav** [comment, marked as answer]
I've played a little bit and managed to make it work. I must say you did an awesome job.

What I did was:
* I have untared all the files myself on the sdcard
* I noticed that `httpd` is actually a busybox binary with all the necessary options, so I have copied and named it `busybox` (in `bin`)
* I have edited most `cgi-bin/*.sh` scripts so 
  * they start with `PATH=/mnt/sdcard/bin:$PATH`
  * shell command names are prefixed with `busybox`. For example `lib=$(dirname "$0")` becomes `lib=$(busybox dirname "$0")`

And that pretty much solves my problem.

Also I figured that `/mnt/mtd/vg_boot.sh` is not executed during the startup sequence of my camera. Instead I have placed the startup script call at the bottom of `/mnt/mtd/nvipcstart.sh`, which works on my cam.

**Reply by Arkady23** [reply]
It's great that you have fully explored new types of cameras. I checked on my camera. Me too nvipcstart.sh used. I will soon release a new version with nvipcstart.sh and without tar. I started using tar because one person had viruses under Windows and they (viruses) changed the size of Linux executable files. When I started responding to him, I noticed that the size of his files was changing. So I started doing the installation from tar. Now there are people who don't have tar in their cell. I will make a new version tomorrow that will install without tar and use nvipcstart.sh . In this case, you will need to make sure that the file sizes do not change when copying. Thank you so much for your research!!!

**Comment by dotvav** [comment]
Closing this discussion, as I feel like I've got what I need. I don't know how frequent my use case is. Feel free to consider this an idea to make your work available to (a few) more.
